File: scala/leetcode/q1/Solution.scala
package leetcode.q1

import scala.collection.mutable

class Solution {}

object Solution {

  def main(args: Array[String]): Unit = {
    val nums = Array(2, 7, 11, 15)
    val target = 9
    val result = twoSum(nums, target)
    print(result.mkString("[", ",", "]"))
  }

  /**
   * @date: 2020/5/31
   * @author: chenguang
   * @desc: 利用hash/map的特性来对数组进行处理，key=num，value=index
   */
  /**
   * @date: 2020/5/31
   * @result: true
   * @time: 70%
   * @memory usage: 100%
   */
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    val map = mutable.Map[Int, Int]()
    for (i <- nums.indices) {
      val num = nums(i)
      if (map.contains(target - num)) {
        return Array(i, map(target - num))
      }
      map.put(num, i)
    }
    Array()
  }

}

File: java/leetcode/q1/Solution.java
package leetcode.q1;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Solution {

    public static void main(String[] args) {
        int[] nums = new int[]{2, 7, 11, 15};
        int target = 9;
        Solution solution = new Solution();
        System.out.println(Arrays.toString(solution.twoSum(nums, target)));
    }

    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i])) {
                return new int[]{map.get(target - nums[i]), i};
            } else {
                map.put(nums[i], i);
            }
        }
        return new int[2];
    }

}
